app/server.ts

```diff
@@ -1,33 +1,59 @@
+import crypto from "crypto";
 import { createServer } from "node:http";
 import next from "next";
 import { Server } from "socket.io";
-import type { ClientToServerEvents, ServerToClientEvents } from "@/types";
+import type { SocketServer } from "@/types";
 import { Field } from "@/utils/game";
 
 const dev = process.env.NODE_ENV !== "production";
 const hostname = "localhost";
 const port = 3000;
+const sessions = new Map<string, "alive" | "dead">();
 
 async function main() {
   const app = next({ dev, hostname, port });
   await app.prepare();
 
   const httpServer = createServer(app.getRequestHandler());
-  const io = new Server<ClientToServerEvents, ServerToClientEvents>(httpServer);
+  const io: SocketServer = new Server(httpServer);
 
   let clientsCount = 0;
 
   let field = await Field.fromRedis();
   field = await field.handleComplete();
 
+  io.use(async (socket, next) => {
+    const { sessionID } = socket.handshake.auth;
+    if (sessionID && sessions.has(sessionID)) {
+      if (sessions.get(sessionID) === "dead") {
+        return next(new Error("dead"));
+      }
+      socket.data.sessionID = sessionID;
+      return next();
+    }
+
+    socket.data.sessionID = crypto.randomBytes(8).toString("hex");
+    next();
```

Comment:
this can be simplified a bit. maybe this is a little clearer also
```suggestion
    const { sessionId } = socket.handshake.auth;
    const sessionState = sessions.get(sessionId);
    if (sessionState === "dead") {
      return next(new Error("dead"));
    }
    socket.data.sessionId = sessionState === "alive"
      ? sessionId
      : crypto.randomBytes(8).toString("hex");
    next();
```